### ISSUE TITLE
Consider unifying approach to PRNG state

### ISSUE THREAD

**Issue by maxentile** [proteneer/timemachine#980]
Thanks to @mcwitt for thoughtful comments: migrating from https://github.com/proteneer/timemachine/pull/978#discussion_r1140414877 . Would be good to discuss and adopt project-wide conventions, if possible.

Some approaches currently used:
* some objects store and update their own random state (integrator, barostat, context, etc.)
* some objects have random functions that accept numpy rng's, jax rng key's, or integer random seeds
* some sites reference global numpy random state

Some possible trade-offs:
* ease of making application code deterministic
* ease of making certain kinds of mistakes
* explicitness
* simplicity

See also: https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html

**Comment by mcwitt** [comment]
Attempting to map out a decision tree:

1. Nondeterministic functions accept PRNG state. This is the approach taken by JAX.
   1. Use numpy API (`numpy.random.Generator`)
   1. Use JAX API (`jax.random.PRNGKey`)
1. Instances with nondeterministic methods maintain their own PRNG state. Seed or initial PRNG state is passed to the constructor.
1. Use global PRNG state. This is the approach taken by the original `numpy.random` API.

**Comment by maxentile** [comment]
Thanks for mapping these out.

By default I lean towards (2) due to familiarity, imposing looser requirements (maybe one class uses cuRAND, a different class only uses numpy, ...), and since it seems harder to make certain kinds of errors (forgetting to split keys etc.). But (1) does seem cleaner.

An additional practice that may be compatible with all of the above options is to implement a random function `random_f(x)` by composing a deterministic function `f(x, gaussian_noise)` (whose implementation is RNG-agnostic) with random generation `gaussian_noise = rng.normal(0, 1, x.shape)` or `gaussian_noise = np.random.randn(*x.shape)` or ...

**Comment by mcwitt** [comment]
> and since it seems harder to make certain kinds of errors (forgetting to split keys etc.)

That's a good point that forgetting to split keys would be a class of error unique to (1b) (added to "disadvantages" above).

**Comment by maxentile** [comment]
> It seems like it does introduce some additional room for error, though. E.g. the caller must be careful to ensure that the input randomness has the expected distribution, and because generation is decoupled from transformation, it might be harder to keep in sync.

That's true, and this is also a realistic concern in the context of reweighting, where we might have a deterministic function `f(theta, samples_from_theta_0)`, expecting input randomness that may be very complicated, expensive, failure-prone to generate.

**Comment by maxentile** [comment]
So it doesn't get lost, some further observations from @mcwitt in : https://github.com/proteneer/timemachine/pull/1128#discussion_r1317571529
